kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/KonanBCEForLoopBodyTransformer.kt

```diff
@@ -0,0 +1,244 @@
+                                if (untilFunction?.symbol?.owner?.name?.asString() == "until" && untilFunction.extensionReceiver?.compareIntegerNumericConst { it >= 0 } == true) {
+                                    val last = untilFunction.getValueArgument(0)!!
+                                    if (last is IrCall) {
+                                        analysisResult = lessThanSize(last)
+                                        // `isLastInclusive` for current case is set to true.
+                                        // This case isn't fully optimized in ForLoopsLowering.
+                                        if (last.isGetSizeCall()) {
+                                            analysisResult = BoundsCheckAnalysisResult(true, (last.dispatchReceiver as? IrGetValue)?.symbol)
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                }
+
+            is WithIndexLoopHeader ->
+                when(loopHeader.nestedLoopHeader) {
+                    is IndexedGetLoopHeader -> {
+                        analysisResult = BoundsCheckAnalysisResult(true,
+                                ((loopHeader.loopInitStatements[0] as? IrVariable)?.initializer as? IrGetValue)?.symbol)
+                    }
+                    is ProgressionLoopHeader -> analysisResult = analyzeLoopHeader(loopHeader.nestedLoopHeader)
+                }
+        }
+        return analysisResult
+    }
+
+    private fun replaceOperators(expression: IrCall, index: IrExpression, safeIndexVariables: List<IrVariable>): IrExpression {
+        if (index is IrGetValue && index.symbol.owner in safeIndexVariables) {
+            val operatorWithoutBC = expression.dispatchReceiver!!.type.getClass()!!.functions.singleOrNull {
+                if (expression.symbol.owner.name == OperatorNameConventions.SET)
+                    it.name == KonanNameConventions.setWithoutBC
+                else
+                    it.name == KonanNameConventions.getWithoutBC
+            } ?: return expression
+            return IrCallImpl(
+                    expression.startOffset, expression.endOffset, expression.type, operatorWithoutBC.symbol,
+                    typeArgumentsCount = expression.typeArgumentsCount,
+                    valueArgumentsCount = expression.valueArgumentsCount).apply {
+                dispatchReceiver = expression.dispatchReceiver
+                for (argIndex in 0 until expression.valueArgumentsCount) {
+                    putValueArgument(argIndex, expression.getValueArgument(argIndex))
+                }
+            }
+        }
+        return expression
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        if (!analysisResult.boundsAreSafe || analysisResult.arrayInLoop == null)
+            return expression
```

Comment:
All the `return expression` in this function should be `return super.visitCall(expression)` in case there is a optimizable array get/set in the call arguments. Please also add some kind of test if possible.